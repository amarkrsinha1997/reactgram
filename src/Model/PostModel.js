import { ValidationError } from '../Execptions/validation-error';
import { NotFoundError } from '../Execptions/not-found';
function PostModel(post) {
  /**
   *  
    post {
      "code": "_A2r0aQcfD",
      "caption": "Some caption",
      "likes": 57,
      "id": "1135147611821557699",
      "display_src": "12276809_750065668431999_184252508_n.jpg",
      "comment": [
        {
          "text": "Uhu!",
          "user": "lucascaixeta",
          "id": 1
        }
      ]
    }
   */
  let _post = post;
  const validateCommentOrThrow = (comment) => {
    if(typeof comment === 'object' && !comment.text) {
      throw new ValidationError('No `comment.text` found in ' + JSON.stringify(comment))
    } else if(!comment) {
      throw new ValidationError('Comment can\'t be empty.')
    }
  };
  this.addComment = (comment) => {
    validateCommentOrThrow(comment)
    _post.comment.push(
      { text: comment.text, user: comment.user ||  'You', id: _post.comment.length + 1 }
    )
  }
  this.deleteComment = (commentId) => {
    const [, commentIndex] = this.findComment(commentId);
    _post.comment = _post.comment.slice(0, commentIndex)
      .concat(_post.comment.slice(commentIndex + 1, _post.comment.length))
  }
  this.updateComment = (commentId, message) => {
    validateCommentOrThrow(message);
    const [comment, commentIndex] = this.findComment(commentId);
    _post.comment[commentIndex] = {
      ...comment,
      text: message
    };
  }
  this.findComment = (commentId) => {
    const commentIndex = _post.comment.findIndex(
      comment => comment.id === commentId
    );
    
    if(commentIndex < 0){
      const message = `No comment for the Id: ${commentId} and postId: ${this.id()}`; 
      throw new NotFoundError(message)
    }
    
    return [_post.comment[commentIndex], commentIndex];
  }
  this.comment = () => _post.comment || [];
  this.post = () => _post;
  this.id = () => _post.id;
}
PostModel.count = 1
PostModel.findPost = (posts, id) => {
  const postIndex = posts.findIndex((post) => post.id() === id);
  if(postIndex < 0) {
    throw new NotFoundError('No post found for Id: ' + id)
  }
  return [posts[postIndex], postIndex]
}
PostModel.executeFor = (posts, postId, cb) => {
  const _posts = posts.slice();
  const [post, postIndex] = PostModel.findPost(posts, postId);
  cb(post);
  _posts[postIndex] = post;
  return _posts;
}
PostModel.addComment = (posts, postId, comment) => {
  return PostModel.executeFor(posts, postId, (post) => post.addComment(comment));
}

PostModel.removeComment = (posts, postId, commentId) => {
  return PostModel.executeFor(posts, postId, (post) => post.deleteComment(commentId));
}

PostModel.updateComment = (posts, postId, commentId, commentMessage) => {
  return PostModel.executeFor(posts, postId, (post) => post.updateComment(commentId, commentMessage));
}

export {PostModel}