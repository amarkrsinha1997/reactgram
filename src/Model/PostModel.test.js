import { PostModel } from './PostModel';
import { ValidationError } from '../Execptions/validation-error';
import { NotFoundError } from '../Execptions/not-found';

describe('PostModel', () => {
  let posts;
  beforeEach(() => {
    posts = [
      new PostModel({
        "code": "_A2r0aQcfD",
        "caption": "Some caption",
        "likes": 57,
        "id": "1135147611821557699",
        "display_src": "12276809_750065668431999_184252508_n.jpg",
        "comment": [
          {
            "text": "Uhu!",
            "user": "lucascaixeta",
            "id": 1
          }
        ]
      }),
      new PostModel({
        "code": "_A2r0aQcfD",
        "caption": "Some caption",
        "likes": 57,
        "id": "1135147611821557698",
        "display_src": "12276809_750065668431999_184252508_n.jpg",
        "comment": [
          {
            "text": "Uhu!",
            "user": "lucascaixeta",
            "id": 1
          }
        ]
      }),
    ]
  })

  it('should be able to add comment', () => {
    const newComment = {
      "text": "Uhu!",
    }
    const post = posts[0];
    PostModel.addComment(posts, post.id(), newComment)
    expect(post.comment()[1]).toStrictEqual({
      ...newComment, "user": "You", id: 2
    })
  });
  it('should be able to edit comment', () => {
    const newComment = {
      "text": "Uhu!",
    }
    const post = posts[0];
    PostModel.addComment(posts, post.id(), newComment);
    const updatedComment = {
      ...newComment, "user": "You", "text": "Uhum!"
    }
    PostModel.updateComment(posts, post.id(), post.comment()[1].id, updatedComment.text)
    expect(post.comment()[1]).toStrictEqual({
      ...updatedComment, id: 2
    })
  });
  it('should be able to delete comment', () => {
    const post = posts[0];
    const comment = post.comment()[0];
    PostModel.removeComment(posts, post.id(), comment.id)
    expect(post.comment()).toStrictEqual([])
  });
  it('should be able to add comment user name in post model if provided.', () => {
    const newComment = {
      "text": "Uhu!",
      "user": "AMar"
    }
    const post = posts[0];
    PostModel.addComment(posts, post.id(), newComment)
    expect(post.comment()[1]).toStrictEqual({
      ...newComment, id: 2
    })
  });
  it('should validate comment when added as yours', () => {
    const newComment = {
      "texts": "Uhu!"
    }
    const post = posts[0];

    const action = () => PostModel.addComment(posts, post.id(), newComment);

    expect(action).toThrowErrorMatchingSnapshot()
    expect(action).toThrowError(ValidationError);
  });
  it('should validate comment when edited.', () => {
    const post = posts[0];

    const action = () => PostModel.updateComment(
      posts,
      post.id(),
      post.comment()[0].id
    )

    expect(action).toThrowErrorMatchingSnapshot()
    expect(action).toThrowError(ValidationError);
  });

  it('should throw error when comment is not found', () => {
    const postId = posts[0].id();
    const commentId = 1223;

    const action = () => PostModel.removeComment(posts, postId, commentId)

    expect(action).toThrowError(NotFoundError);
    expect(action).toThrowErrorMatchingSnapshot();
  });
  it('should throw error when post is not found', () => {
    const postId = '44343232232321';
    
    const action = () => PostModel.findPost(posts, postId);

    expect(action).toThrowError(NotFoundError);
    expect(action).toThrowErrorMatchingSnapshot();
  });

  it('should be able to find the comment', () => {
    const index = 0;
    const post = posts[index];
    const comment = post.comment()[index]
    const result = post.findComment(comment.id);
    expect(result).toStrictEqual(
      [comment, index]
    );
  });
  it('should be able to find the post', () => {
    const index = 0;
    const post = posts[index];
    expect(PostModel.findPost(posts, post.id())).toStrictEqual([posts[index], index]);
  });
});