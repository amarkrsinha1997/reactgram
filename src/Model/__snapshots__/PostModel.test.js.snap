// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostModel should throw error when comment is not found 1`] = `"No comment for the Id: 1223 and postId: 1135147611821557699"`;

exports[`PostModel should throw error when post is not found 1`] = `"No post found for Id: 44343232232321"`;

exports[`PostModel should validate comment when added as yours 1`] = `"No \`comment.text\` found in {\\"texts\\":\\"Uhu!\\"}"`;

exports[`PostModel should validate comment when edited. 1`] = `"Comment can't be empty."`;
